### DISCUSSION TITLE
Updating Prisma causes type errors in interactive transactions

### DISCUSSION THREAD

**Discussion by Bohemus307** [prisma/prisma#18126] [Q&A]
### Bug description

I updated from 4.6.1 to 4.7.0 and now I am receiving a type-error around the use of interactive transactions no longer as a preview.  The error message is around the promise argument and no amount of upgrading downgrading and prisma generate seems to solve this issue. I am using Nestjs am receiving this error.

`src/services/webhook-actions/webhook-actions.service.ts:771:37 - error TS2769: No overload matches this call.
  Overload 1 of 2, '(arg: PrismaPromise<any>[], options?: { isolationLevel?: TransactionIsolationLevel; }): Promise<any[]>', gave the following error.
    Argument of type '(prisma: PrismaClient) => Promise<Transaction>' is not assignable to parameter of type 'PrismaPromise<any>[]'.
  Overload 2 of 2, '(fn: (prisma: Omit<PrismaService, "$connect" | "$disconnect" | "$on" | "$transaction" | "$use">) => Promise<Transaction>, options?: { ...; }): Promise<...>', gave the following error.
    Argument of type '(prisma: PrismaClient) => Promise<Transaction>' is not assignable to parameter of type '(prisma: Omit<PrismaService, "$connect" | "$disconnect" | "$on" | "$transaction" | "$use">) => Promise<Transaction>'.
      Types of parameters 'prisma' and 'prisma' are incompatible.
        Type 'Omit<PrismaService, "$connect" | "$disconnect" | "$on" | "$transaction" | "$use">' is not assignable to type 'PrismaClient<PrismaClientOptions, never, RejectOnNotFound | RejectPerOperation>'.

771     return this.prisma.$transaction(async (prisma: PrismaClient) => {
                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~`


Heres is a example component where the error is occuring.

`/**
   * Create endpoints
   */
  @Post('/api/v1/add-user')
  @HttpCode(200)
  @ApiResponse({
    status: 200,
    description: 'Success',
    type: AddUserV1Response,
  })
  @ApiAuth()
  async create(@Req() request: Request): Promise<any> {
    return this.honeycombService.start(request, {}, async () => {
      const body = new AddUserRequest({
        firstName: request.body.firstName,
        lastName: request.body.lastName,
        email: request.body.email,
        remoteId: request.body.remoteId,
        countryCode: request.body.countryCode,
        preferredCurrency: request.body.preferredCurrency,
        isAdmin: request.body.isAdmin ?? false,
      });

      await validateRemapped(body, () => {
        this.honeycombService.log(HoneycombEvents.CREATE_USER_ERROR, {
          [HoneycombKeys.ERROR_CAUSE]: 'form',
          [HoneycombKeys.PAYLOAD]: JSON.stringify(request.body).slice(0, 1000),
        });
      });

      return this.authService.authorize(
        AdminPermission.UserWrite,
        request,
        async (corpReq) => {
          const maybeCountry = await this.getCountry(body.countryCode);
          const corporationId = corpReq.corporationId;

          return this.prisma.$transaction(
            async (prisma: PrismaClient) => {
              const { newUser, grantEmployee, d4dEmployee } =
                await this.userService.createUser(prisma, corporationId, {
                  email: body.email,
                  password: null,
                  firstName: body.firstName,
                  lastName: body.lastName,
                  remoteId: body.remoteId,
                  country: maybeCountry,
                  preferredCurrency: body.preferredCurrency,
                  isAdmin: body.isAdmin,
                });

              return {
                userId: newUser.user_id,
                donorAccountNumber: newUser.account_number,
                employeeAccountNumber: grantEmployee.account_number,
                d4dEmployeeAccountNumber: d4dEmployee.account_number,
              };
            },
            { timeout: 20000 },
          );
        },
      );
    });
  }`

Ive been digging at this for days and have seen a lot of similar issues around github etc. but cannot find a possible solution to this bug and would love any help i can get. 

### How to reproduce

<!--
1. Go to '...'
2. Change '....'
3. Run '....'
4. See error
-->


### Expected behavior

Expected either the transactions to still work as they did with preview or migration documentation for andy issues that arose with the change?

### Prisma information

<!-- Do not include your database credentials when sharing your Prisma schema! -->
```prisma
// Add your schema.prisma
```

```ts
// Add your code using Prisma Client
```


### Environment & setup

- OS: Mac Os
- Database: PostgresQL
- Node.js version: Node 18.13.0


### Prisma Version

```
4.10.1
```

**Comment by nurul3101** [comment]
Hey @Bohemus307 👋 

Does the error go away if you downgrade to version 4.6.1?

**Reply by Bohemus307** [reply]
@nurul3101  Unfortunately it does not. No amount of rollback so far has helped but today i might try and go back to 3.11 where the code was before i began trying to help it out!

**Reply by nurul3101** [reply]
If I understand correctly, you were on version 3.11.0 and updated to 4.6.1 without any issues. Then updating from 4.6.1 to 4.7.0 caused this issue?

**Reply by Bohemus307** [reply]
actually it was just the one upgrade from 3.11 to 4.7 but i tried to go baclk to 4.6.1 to solve the issue. What i found yesterday was my npx version of prisma somehow got set to 4.11 and was causing the issue down grading back to 4.7 has removed my typescript error. Thanks for any help on this.

**Comment by jessebs** [comment]
I just started experiencing this same bug when upgrading from 5.12.1 -> 5.17.0.  It looks like it was introduced in 5.16.0 as rolling back to 5.15.1 makes it go away.
